### PR TITLE
Add support for directly pasting story url

### PR DESCRIPTION
**Pull request by AaronBenDaniel**
The frontend still says that it needs to accept a story ID and not a full link, but I didn't want to change that because you probably have your own preferences as to how it should be changed.

### CODE REVIEW

File: src/api/src/main.py [hunk 1]
```diff
@@ -26,11 +33,14 @@ def home():
 
 @app.get("/download/{story_id}")
 async def download_book(
-    story_id: int,
+    story_id: str,
```
Comment (TheOnlyWayUp) [inline src/api/src/main.py:36]: Let's make this `story_url` and add another `story_id` parameter. Only one is required.

Story URL would be reflected on the front end, and story id is present for easier integration with code and for backward compatibility.

File: src/api/src/main.py [hunk 2]
```diff
@@ -26,11 +33,14 @@ def home():
...
     download_images: bool = False,
     username: Optional[str] = None,
     password: Optional[str] = None,
 ):
+
+    story_id=extract_story_ID(story_id)
```
Comment (TheOnlyWayUp) [inline src/api/src/main.py:42]: The logic is simple enough to not warrant placing this code in a function.

I'm yet to write a contributions.md, but I'd appreciate if you used `black` to format your code!

File: src/frontend/src/routes/+page.svelte [hunk 3]
```diff
@@ -16,7 +16,7 @@
     (is_paid_story && !(credentials.username && credentials.password));
 
   $: url =
-    `/download/${story_id}?om=1` +
+    `/download/${story_id.replaceAll("/","")}?om=1` +
```
Comment (TheOnlyWayUp) [inline src/frontend/src/routes/+page.svelte:19]: Actually, let's change _this_ (param on the frontend) to Story URL, and do all the extraction through javascript.

This way, the API stays unchanged.

File: src/api/src/main.py (hunk 1 above)
Comment (TheOnlyWayUp) [reply src/api/src/main.py:36]: Oops, scratch this!

Please refer to https://github.com/TheOnlyWayUp/WattpadDownloader/pull/8#discussion_r1667403451

**Review by TheOnlyWayUp — CHANGES_REQUESTED** [review]
Let's make this a frontend-exclusive change.

We'll rename Story ID on the frontend to Story URL, and do the extraction through Javascript. We also need to maintain support for just Story IDs on the frontend.

The 'How do I find a Story ID?' Help text would need to be updated as well.

**Comment by AaronBenDaniel** [comment]
I have not changed the help modal since adding support for Part IDs and URLs. Honestly, I don't think it's even necessary anymore, but I want your input.

**Comment by AaronBenDaniel** [comment]
After some more testing, I have found a flaw that I think might be impossible to fix. If the user gives a part ID for a book that is a story ID for another book, it will attempt to downloader the incorrect book.

**Comment by AaronBenDaniel** [comment]
We could work around it by removing support for giving only Part and Story IDs and requiring users to give full URLs. This way, we'll know if the number came from a Part URL or a Story URL and we can send that information to the API so it can get the right book.

I don't love this solution, as it removes functionality, but I think it's worth the tradeoff.

File: src/api/src/create_book.py [hunk 4]
```diff
@@ -143,6 +143,39 @@ async def fetch_cover(url: str, cookies: Optional[dict] = None) -> bytes:
...
+@backoff.on_exception(backoff.expo, ClientResponseError, max_time=15)
+async def fetch_story_ID(part_id: int, cookies: Optional[dict] = None) -> int:
+    """Return the Story ID given a Part ID"""
+    async with (
+        CachedSession(headers=headers, cache=cache)
+        if not cookies
+        else ClientSession(headers=headers, cookies=cookies)
+    ) as session:  # Don't cache requests with Cookies.
+        async with session.get(
+            f"https://www.wattpad.com/api/v3/story_parts/{part_id}?fields=url"
```
Comment (TheOnlyWayUp) [inline src/api/src/create_book.py:156]: @AaronBenDaniel Good stuff! How did you find this endpoint?

Comment (AaronBenDaniel) [reply src/api/src/create_book.py:156]: I found the one for getting a Part URL from a Part ID in wattpad-py. In my searching I found the same method in wattpad-rs as well. The "endpoint" for getting a Story URL from a Part URL is just downloading the HTML content of the Part URL.

I asked myself, "how does a human find the Story URL from a Part URL" and realized that it's just in the top left-hand corner of the reading interface. That URL appears to be gotten directly from the Part URL and not the API, so the HTML content of the web reading interface contains the Story URL. This just fetches that content and extracts the Story URL.

**Comment by TheOnlyWayUp** [comment]
@AaronBenDaniel Take a look at https://github.com/TheOnlyWayUp/WattpadDownloader/blob/e16496ca94f6c04c3569c0e5984d93b7de854859/src/frontend/src/routes/%2Bpage.svelte#L20-L44

Thanks for sharing that `story_parts` API Endpoint! It would be cool if we could derive the Story ID from a pasted Part URL on the Frontend.

**Pull request closed without merging**